include/RAJA/policy/hip/atomic.hpp

```diff
@@ -188,327 +68,684 @@ using hip_atomicCommon_builtin_types = list<
      ,unsigned long long
     >;
 
-using hip_atomicAdd_builtin_types = list<
-      int
-     ,unsigned int
-     ,unsigned long long
-     ,float
-#ifdef RAJA_ENABLE_HIP_DOUBLE_ATOMICADD
-     ,double
-#endif
-    >;
 
-/*!
- * List of types where HIP builtin atomics are used to implement atomicSub.
+/*
+ * Performs an atomic exchange. Stores the new value in the given address
+ * and returns the old value.
  */
-using hip_atomicSub_types = list<
-      int
-     ,unsigned int
-     ,float
-#ifdef RAJA_ENABLE_HIP_DOUBLE_ATOMICADD
-     ,double
-#endif
-    >;
+template <typename T,
+          std::enable_if_t<std::is_same<T, int>::value ||
+                           std::is_same<T, unsigned int>::value ||
+                           std::is_same<T, unsigned long long int>::value ||
+                           std::is_same<T, float>::value, bool> = true>
+RAJA_INLINE __device__ T hip_atomicExchange(T *acc, T value)
+{
+  return ::atomicExch(acc, value);
+}
+
+template <typename T,
+          std::enable_if_t<!std::is_same<T, int>::value &&
+                           !std::is_same<T, unsigned int>::value &&
+                           !std::is_same<T, unsigned long long int>::value &&
+                           !std::is_same<T, float>::value &&
+                           sizeof(T) == sizeof(unsigned int), bool> = true>
+RAJA_INLINE __device__ T hip_atomicExchange(T *acc, T value)
+{
+  return RAJA::util::reinterp_A_as_B<unsigned int, T>(
+    hip_atomicExchange(reinterpret_cast<unsigned int*>(acc),
+                       reinterpret_cast<unsigned int&>(value)));
```

Comment:
We should probably reinterpret the bits from the old value into a new value instead of changing the reference type to avoid undefined behavior where possible.

Reply:
Can you provide an example of what you mean? The second argument is taken by value, which is then passed to the underlying ::atomicExch function.

Reply:
I mean if its a `float value;` then you do `reinterpret_cast<unsigned int&>(value)` you make a reference bound to a value of a different type which is undefined behavior.

Reply:
To get well defined bahavior you can memcpy between a float and an unsigned int which is what RAJA::util::reinterp_A_as_B should be doing.

Reply:
Oh, do you mean like a memcpy of the bytes of value into a stack allocated unsigned int?

Reply:
Done.